src/Symfony/Component/String/Utf8String.php

```
@@ -0,0 +1,261 @@
+        }
+
+        return $string === $this->string;
+    }
+
+    public function indexOf($needle, int $offset = 0): ?int
+    {
+        if ($needle instanceof AbstractString) {
+            $needle = $needle->string;
+        } elseif (\is_array($needle) || $needle instanceof \Traversable) {
+            return parent::indexOf($needle, $offset);
+        } else {
+            $needle = (string) $needle;
+        }
+
+        if ('' === $needle) {
+            return null;
+        }
+
+        $i = $this->ignoreCase ? mb_stripos($this->string, $needle, $offset, 'UTF-8') : mb_strpos($this->string, $needle, $offset, 'UTF-8');
+
+        return false === $i ? null : $i;
+    }
+
+    public function indexOfLast($needle, int $offset = 0): ?int
+    {
+        if ($needle instanceof AbstractString) {
+            $needle = $needle->string;
+        } elseif (\is_array($needle) || $needle instanceof \Traversable) {
+            return parent::indexOfLast($needle, $offset);
+        } else {
+            $needle = (string) $needle;
+        }
+
+        if ('' === $needle) {
+            return null;
+        }
+
+        $i = $this->ignoreCase ? mb_strripos($this->string, $needle, $offset, 'UTF-8') : mb_strrpos($this->string, $needle, $offset, 'UTF-8');
+
+        return false === $i ? null : $i;
+    }
+
+    public function length(): int
+    {
+        return mb_strlen($this->string, 'UTF-8');
+    }
+
+    public function prepend(string ...$prefix): AbstractString
+    {
+        $str = clone $this;
+        $str->string = (1 >= \count($prefix) ? ($prefix[0] ?? '') : implode('', $prefix)).$this->string;
+
+        if (!preg_match('//u', $str->string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        return $str;
+    }
+
+    public function replace(string $from, string $to): AbstractString
+    {
+        $str = clone $this;
+
+        if ('' === $from || !preg_match('//u', $from)) {
```

Wouldn't it make sense to throw herein the same way as we do it for `$to`?

---

If from `$from` is not UTF-8, it won't be found, thus won't be replaced, that's all. For `$to` its a different story: the result must be UTF-8, so it has to be too.

---

Understood, but still doesn't that mean that there is a mistake on the user's side that could more easily be detected when we throw here instead of silently ignore it?

---

That's against the design of the component, e.g. `indexOf()` and many more methods would have to throw to be consistent with this policy. Yet I think that would make the component harder to use, forcing ppl to add try/catch all around. Existing string functions (especially `mb_*`) don't fail hard and that's great. We should follow their lead.